zerver/lib/message.py

```diff
@@ -212,9 +212,16 @@ def messages_for_ids(
         if message_id in search_fields:
             msg_dict.update(search_fields[message_id])
         # Make sure that we never send message edit history to clients
-        # in realms with allow_edit_history disabled.
-        if "edit_history" in msg_dict and not allow_edit_history:
-            del msg_dict["edit_history"]
+        # in realms according to show_message_edit_history setting.
+        if "edit_history" in msg_dict:
+            if show_message_edit_history == 2:
+                for history in msg_dict["edit_history"][:]:
+                    if "prev_content" in history:
+                        msg_dict["edit_history"].remove(history)
+                if len(msg_dict["edit_history"]) == 0:
+                    del msg_dict["edit_history"]
+            elif show_message_edit_history == 3:
+                del msg_dict["edit_history"]
```

Comment:
Ideally, we'd arrange that this call the same function that I suggested adding in the views code, so we're not implementing this logic twice. (That oftens leads to bugs).  You may need to fiddle with the interface to make it work well.